### ISSUE TITLE
Modular/separate deployment for agones-allocator

### ISSUE THREAD

**Issue by devloop0** [googleforgames/agones#1691]
**Is your feature request related to a problem? Please describe.**
The agones-allocator can function completely separately from the rest of the components (controller, ping, etc.). It would be nice to refactor that into its own helm deployment. 
One usecase for a separate agones-allocator would be as a service that a matchmaker could call out to (so the matchmaker wouldn't also need to be able to host gameservers).

**Describe the solution you'd like**
Ideally, we'd be able to refactor to be more modular. As an example, I was able to deploy agones-allocator with just allocation.yaml file and the crd definitions (though in reality, we'd probably only need the CRDs for gameserver allocation policies and gameservers is my guess).
Even better would be if there were some helm option that could just install certain components (not sure if this is possible).

**Describe alternatives you've considered**
A user could deploy the whole agones installation and just use agones-allocator, but this would be wasteful.

**Comment by roberthbailey** [comment]
> The agones-allocator can function completely separately from the rest of the components (controller, ping, etc.)

As you mentioned, it still requires (at least some of) the CRDs to be installed as well to function. Otherwise, it would be as simple as just applying a deployment yaml to a cluster. 

> Even better would be if there were some helm option that could just install certain components

I think we could do this by adding a bunch of if/else statements to the helm chart.... I was trying to figure out if there is a way in helm to compose charts together in a reasonable way (so you could create a stand-alone chart for the allocator and include that into the overall chart) but I haven't found anything promising yet.

**Comment by markmandel** [comment]
We do already have options to remove items from the helm chart installation -- maybe the answer is to add more of these?

I.e. have more negative removal, than a positive standalone aspect? (if that makes sense?)

**Comment by roberthbailey** [comment]
I think we are saying the same thing. Something like:

- CRDs are always installed (no conditionals)
- Agones controller (and related stuff) is installed by default and has a conditional to skip it
- Allocator service is installed by default and has a conditional to skip it
- Stackdriver support is _not_ installed by default and has a conditional to install it
- etc.

This way if you do nothing special, you get the default, it just works, configuration. But if you want to install just part of agones, then you can turn off the bits that aren't necessary in your cluster.

**Comment by markmandel** [comment]
100% in agreement.

**Comment by devloop0** [comment]
It seems like a lot of stuff already is behind an agones.*.install. From what I could see, the only exception is the controller, so maybe adding an agones.controller.install would be sufficient to solve *this* issue.

Just my two cents, but I think the existing agones.*.install specifically for services (allocator, etc.), doesn't really convey what a user might want. As an example, if I set agones.allocator.install, I would expect that the allocator and everything related (like CRDs) would be installed. So if the user naively goes with setting every agones.*.install to false except agones.allocator.install, they'll also set agones.crds.install to false, which means they'll have a non-functioning allocator installed (this is what happened to me initially). 

Ideally, there would be an agones.*.install for each specific service/module such that setting one of them to true will set every related thing that it needs to also be true. As an example, let's say services A and B both need CRDs and agones.A.install is true but agones.B.install is false. The in my vision, agones.A.install will not only install not only A's resources (pods, services, secrets, etc.), but also any external resources it needs as well (e.g. crds). So there would be some core (small) subset of flags such that if all of them are set to false, then nothing is installed, and then setting stuff to true will only install the minimal set of things that are needed (and all of them set to true will be the default with everything added).

**Comment by roberthbailey** [comment]
Are there any cases where you would use the agones helm installer and not need the CRDs?

**Comment by devloop0** [comment]
Maybe it's to keep an old version of CRDs even if you upgrade to newer versions of services? Not 100% sure.

**Comment by markmandel** [comment]
> Are there any cases where you would use the agones helm installer and not need the CRDs?

There used to be issues with installing Agones as a sub-chart because of CRD installation ordering, and therefore you would have to remove CRDs from the Helm chart, and do a copy-paste operation to move it into the parent chart. I'm not 100% sure if this is a solved issue with Helm3, but it does seem to be - so we have the option to tell this chart not to install the CRDs, so you can choose to have another system / chart install them.

I can't find the initial conversation - but there's lots of discussion here: https://github.com/googleforgames/agones/issues/1280

TL;DR - sometimes you want to control your own CRD lifecycle, outside of the Agones Helm Chart.

**Comment by pooneh-m** [comment]
> So if the user naively goes with setting every agones.*.install to false except agones.allocator.install, they'll also set agones.crds.install to false, which means they'll have a non-functioning allocator installed (this is what happened to me initially).

I think it makes sense that GameServerAllocationPolicy is conditioned only on on `agones.allocator.install` or `agones.allocator.crd.install`.

**Comment by roberthbailey** [comment]
This issue hasn't been updated in a while. Is this still something we think would be valuable? We have a lot of configurability in our helm charts already, and I only want to add more if it's something that people actually need and will use.
